package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,335 @@
+         cite_module=True)
+due.cite(Doi("10.1051/sfn/201112010"),
+         description="FCA fast correlation algorithm",
+         path="MDAnalysis.analysis.msd",
+         cite_module=True)
+del Doi
+
+class EinsteinMSD(object):
+    r"""Class to calculate Mean Squared Displacement by the Einstein relation.
+    
+    Attributes
+    ----------
+    dim_fac : float
+        Dimensionality :math:`d` of the MSD.
+    timeseries : :class:`np.ndarray`
+        The averaged MSD with respect to lag-time.
+    msd_per_particle : :class:`np.ndarray`
+        The MSD of each individual particle with respect to lag-time.
+    N_frames : int
+        Number of frames in trajectory.
+    N_particles : int
+        Number of particles MSD was calculated over.
+
+    """
+
+    def __init__(self, u, selection=None, msd_type='xyz', fft=True):
+        r"""
+        Parameters
+        ----------
+        u : Universe
+            An MDAnalysis :class:`Universe`.
+        selection : str
+            An MDAnalysis selection string. Defaults to `None`.
+        msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Desired dimensions to be included in the MSD. Defaults to 'xyz'.
+        fft : bool
+            Use a fast FFT based algorithm for computation of the MSD.
+            Otherwise, use the naieve or "simple" algorithm. Defaults to `True`
+
+        """
+        #args
+        self.u = u
+        self.selection = selection
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self._dim = None
+        self._position_array = None
+        
+        #indexing
+        self.N_frames = len(self.u.trajectory)
+        self.N_particles = 0
+
+        #result
+        self.dim_fac = 0
+        self.timeseries = None
+        self.msd_per_particle = None
+
+        #prep
+        self._prepare()
+
+            
+    def _prepare(self):
+        self._parse_msd_type()
+        self._select_reference_positions()
+        
+    def _parse_msd_type(self):
+        r""" Sets up the desired dimensionality of the MSD.
+        
+        Parameters
+        ----------
+        self.msd_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+            Dimensions to be included in the MSD.
+
+        Returns
+        -------
+        self._dim : list
+            Array-like used to slice the positions to obtain desired dimensionality
+        self.dim_fac : float
+            Dimension factor :math:`d` of the MSD.
+
+        """
+        keys = {'x':0, 'y':1, 'z':2}
+        dim = []
+
+        self.msd_type = self.msd_type.lower()
+        print(self.msd_type)
+        try:
+            for tok in self.msd_type:
+                vals = (keys.pop(tok))
+                dim.append(vals)
+        except KeyError:
+            raise ValueError('invalid msd_type specified, please specify one of xyz, xy, xz, yz, x, y, z ')
+        
+        dim.sort()
+        self._dim = dim
+        self.dim_fac = float(len(dim))
```

Comment:
It never could be outside of precision errors, because keys are popped from the dictionary of length 3. That being said this all seems quite convoluted -- you could just have a dictionary for your 6 cases and sort+lowercase the input.